[MRG] K-SVD

alexeyum opened the pull request:
Implemented a K-SVD algorithm as a free function.

Notes/questions:
- I haven't done any optimizations yet, but I left TODO-notes in places where it can be done. I am going to profile the code later and see where/if the optimizations are necessary.
- The algorithm currently uses OMP as a sparse coder, and it is possible to make it customizable. However this will make the function interface more complicated (as different sparse coders have different parameters). Should I do it, or should I leave it like this?
- There are now no unit-tests for the implementation, but I am a bit confused on how to create them for such functions. Should I just check the API and simple functions like `_worst_represented_example`?

alexeyum committed:
K-SVD free function and example.

alexeyum committed:
Updated the K-SVD algorithm.

Main changes:
* Added a DictionaryLearningKSVD class.
* Merged the K-SVD and DictionaryLearning examples.
* learn_dictionary_ksvd now returns errors path.
* Added tests.

alexeyum committed:
Changed norm calculation for numpy 1.6 compatibility.

alexeyum committed:
Used utils.random.choice for numpy 1.6 compatibility.

alexeyum committed:
Changed norm calculation for numpy 1.6 compatibility (one more place).

alexeyum committed:
Added a benchmark for K-SVD

alexeyum committed:
Removed some forgotten TODO-notes.

arthurmensch commented on sklearn/decomposition/ksvd.py:
```diff
@@ -0,0 +1,343 @@
+"""K-SVD dictionary learning algorithm.
+"""
+
```
Given the structure of your code, I think it would be worth factorizing it it within `dict_learning` module, adding a new `_update_dict(method='ksvd')` method (from your `_update_atom` method and allowing for `method='ksvd' in `dictionary_learning`.

alexeyum replied:
Technically it is not hard to merge the functions completely and make possible every combination of initialization (SVD or random samples), sparse coding (lars, omp, ...) and dictionary updating (CD or K-SVD). Then the problem statement (L1 or L0) and parameters type (alpha or n_nonzero_coefs) depend on the sparse coding method.

However it may seem confusing for users, as some combinations do not correspond to well-known dictionary learning methods. We can still make aliases like `ksvd` that use canonical combinations.

agramfort replied:
+1 for fusing implementations. Then we can see if we need an extra ksvd
function. I think it's not necessary if it's documented. It also helps
users to realize that KSVD is not fundamentally different from other
dictionary learning approaches.

arthurmensch commented on sklearn/decomposition/ksvd.py:
```diff
@@ -0,0 +1,343 @@
+    return norm(errors)
+
+
+def _init_dictionary_from_samples(X, n_components, random_state):
```
In `dict_learning`, we use an svd for finding initial atoms. I think it would be more consistent to allow doing it in k-SVD. Maybe provide two initialization  methods in `dict_learning` module ?

arthurmensch commented on sklearn/decomposition/ksvd.py:
```diff
@@ -0,0 +1,343 @@
+def _atom_update(X, atom_index, dictionary, sparse_codes, approximate_svd):
+    """Update single dictionary atom and the corresponding codes."""
+
+    atom_usages = numpy.nonzero(sparse_codes[:, atom_index])[0]
```
scikit-learn use `np` abbreviation

arthurmensch commented on sklearn/decomposition/ksvd.py:
```diff
@@ -0,0 +1,343 @@
+
+    # Sparse version is used because non-sparse does not allow to
+    # set the desired number of singular values.
+    U, S, V = scipy.sparse.linalg.svds(residual, 1)
```
Is that the most efficient way to do a 1-rank decomposition of a matrix ?

eickenberg replied:
If you need the vector, probably yes - it uses the power method IIRC

On Fri, Aug 28, 2015 at 1:53 PM, Arthur Mensch <notifications@github.com>
wrote:

> In sklearn/decomposition/ksvd.py
> <https://github.com/scikit-learn/scikit-learn/pull/3739#discussion_r38192315>
> :
>
> > +
> > +def _worst_represented_example(X, dictionary, sparse_codes):
> > +    """Find the sample that has the most representation error."""
> > +
> > +    residuals = X - numpy.dot(sparse_codes, dictionary)
> > +    errors_squared = numpy.sum(residuals * residuals, axis=1)
> > +    worst_index = numpy.argmax(errors_squared)
> > +    return X[worst_index, :]
> > +
> > +
> > +def _decompose_svd(residual):
> > +    """Decompose the residual using 1-rank SVD."""
> > +
> > +    # Sparse version is used because non-sparse does not allow to
> > +    # set the desired number of singular values.
> > +    U, S, V = scipy.sparse.linalg.svds(residual, 1)
>
> Is that the most efficient way to do a 1-rank decomposition of a matrix ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3739/files#r38192315>.
>

arthurmensch replied:
Seems so. It also look like you can provide an initial vector approximation, could you try with present atom ?

alexeyum commented:
Hi,

I am now working on fusing implementations and I have several questions:
 1. Is there any paper that describes the current `dict_learning` method that I can refer to? I think it would be good to have references for both K-SVD and L1-based methods in the documentation.
 2. Is it okay if I remove or rename some local variables in `dict_learning`? It may affect applications that use the `callback` argument as it has `locals()` as argument.
 3. In `dict_learning` there is currently an argument called `code_init` which is not actually used (as the first step in alternative iteration is sparse coding). I suggest removing it (`dict_learning_online` does not have similar argument).

arthurmensch commented:
1. You can refer to Dictionary Learning Algorithms for Sparse Representation, Kreutz-Delgado et al. for the `dict_learning` algorithm, Online learning for matrix factorization and sparse coding, Mairal et al for the `dict_learning_online` algorithm.

2. Please avoid changing variable names in `dict_learning` (what change would you want to make) as these are consistent with `dict_learning_online` and the rest of the module. Especially, the API should not be broken if there is no strong reason to do so.

3. Code_init is used as a warm start in coordinate descent sparse coding, as mentionned in the doc. This is not present in `dict_learning_online` as Lasso is performed on mini batches.

alexeyum commented:
1. OK, thanks.

2. I rechecked the code and it seems that I can update the code without changing any variables names. I will tell you if any problems arise.

3. Yes, sorry, found the way it is used in the code.

alexeyum committed:
Partial work for interface change

alexeyum committed:
Merged conflicts with master

alexeyum committed:
Removed merge info

alexeyum committed:
Finished moving K-SVD to dict_learning interface

alexeyum committed:
Added usage of dict_learning interface in K-SVD benchmark

alexeyum committed:
Updated all documentation.

alexeyum committed:
Added tests for different run parameters.

alexeyum commented:
Hi, I merged the new K-SVD code with the old L1-based dictionary learning implementation, and now the PR is ready for second review.

alexeyum commented:
The AppVeyor check failed, though the test doesn't seem to be related to the changed files. Can it be floating?

agramfort commented:
@arthurmensch do you have time to look?

arthurmensch commented:
Yes I have been following, I ll review it in the next few days now that it
is ready
Le 13 juil. 2016 17:41, "Alexandre Gramfort" <notifications@github.com> a
écrit :

> @arthurmensch <https://github.com/arthurmensch> do you have time to look?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3739#issuecomment-232384336>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AD48Y0p4KFS5kZD-kOWCYnEV8ayGhUl-ks5qVP6zgaJpZM4CrJka>
> .
>

alexeyum committed:
Fixed documentation merge conflicts.

alexeyum committed:
Removed accidental documentation formatting change.

alexeyum committed:
Removed redundant blank line.

alexeyum committed:
PEP8 style fixes.

alexeyum commented:
Hi,

I fixed the merge issues, but the PR still needs the second review.

I am not sure, but it seems that the last check fail is not caused by my code - it has just timed out.

I also fixed all the flake8 warnings except E402 in file `examples/decomposition/plot_image_denoising.py`, which is caused by the `print(__doc__)` [(link to the line)](https://github.com/alexeyum/scikit-learn/blob/58a60c7aa31255012801889940c0361e03ec7588/examples/decomposition/plot_image_denoising.py#L37). I figured out I shouldn't fix it as it is present in all the examples.

amueller commented:
Yeah there's a patch to allow the E402 in the examples in #8036 (which I guess is why it doesn't fail any more)

alexeyum commented:
So, is there anything I should fix?

jnothman commented:
@arthurmensch, any chance of still squeezing that review from you? The benchmarks above look great, but I'm inundated with reviews in technical spaces I feel I have competence in, unlike here!

alexeyum commented:
Hi! Maybe I should ping the mailing list and ask for some reviews?

arthurmensch commented:
I'll take the review

agramfort commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -936,20 +1164,37 @@ class DictionaryLearning(BaseEstimator, SparseCodingMixin):
+            error matrix decomposition; in the most cases this
+            method is more preferable than the 'exact_ksvd' as it
+            gives good precision but works faster. The sparse coder
+            for learning is OMP.
```
I think ksvd is a misnomer. It should be omp as lars and cd refer to the algo used for sparse coding.

alexeyum replied:
I agree, 'omp' is better.

alexeyum replied:
Upon consideration I think the name 'omp' is somewhat confusing too as in the sparse coding community K-SVD is a well-known dictionary learning algorithm, but OMP is just a sparse coding algorithm. I think including 'ksvd' in the name might help those searching for this algorithm a lot. What do you think of 'omp_ksvd' or 'ksvd_omp'?

agramfort commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -936,20 +1164,37 @@ class DictionaryLearning(BaseEstimator, SparseCodingMixin):
+            for learning is OMP.
+        exact_ksvd : update atoms separately using exact 1-svd
+            error matrix decomposition. The sparse coder for learning
+            is OMP.
```
do we really need to support both?

alexeyum replied:
AFAIK it is not commonly used, but it was easy to implement, so I did it just in case. I can remove it if you want.

agramfort commented:
please remove what you don't recommend to use

alexeyum committed:
Fixed merge conflicts

alexeyum committed:
Removed the exact_ksvd option

alexeyum committed:
Fixed styleguide

alexeyum committed:
More styleguide fixes

alexeyum commented:
(not sure my comment for outdated code is visible, copying it here)

Upon consideration I think the name 'omp' is somewhat confusing too as in the sparse coding community K-SVD is a well-known dictionary learning algorithm, but OMP is just a sparse coding algorithm. I think including 'ksvd' in the name might help those searching for this algorithm a lot. What do you think of 'omp_ksvd' or 'ksvd_omp

pirofti commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -309,9 +314,109 @@ def sparse_encode(X, dictionary, gram=None, cov=None, algorithm='lasso_lars',
+        dictionary[atom_index, :] = atom / norm(atom)
+    elif len(atom_usages) == 1:
+        # update is trivial if the atom is used in only one sample
+        atom = X[atom_usages[0], :]
```
Not sure this is correct. The new atom is definitely not replaced by the signal using it.  Not in the original algorithm anyway.

I mean, I can see why you would reason that way: it is the only signal using this atom, so why not make the atom be exactly the signal which would lead to a perfect approximation (to equality).

The thing is that we have no guarantee that the atom at this iteration will also be used by a single signal at future iterations. That is to say, if now only one signal uses it, perhaps future changes will push others to also use it. Forcing it to be replaced with the actual signal prohibits that possibility.

Think of it this way, if the current atom is very close to another atom (their mutual coherence is almost 1), then signals will migrate from one to the other during dictionary learning iterations until they grow further apart or until one of them is replaced by the `_worst_represented_example`.

pirofti reviewed (commented):
I hope I am not joining too late in the game, but I would like to see K-SVD available in scikit-learn as it is the de-facto dictionary learning algorithm in the field.

Just finished reviewing the latest changes and I think the modifications proposed by @alexeyum can go in as they are or with the modifications that I proposed. If they go in as is, I can create a pull request later with my minor tweaks.

While Approximate K-SVD is used more often than its exact form, K-SVD, I think it would be proper to have both as options. The first is fast, as it uses a single iteration of the power method to update the dictionary, which makes it very well suited for approximation tasks such as de-noising, while the later represents the optimal solution to the atom update optimization problem which allows it to converge faster and provides us with some nice properties needed by K-SVD-based variations (such as regularization).

Let me know if there is anything I can do to move this along. I am new to the project, but I can speak dictionary learning fluently. I have written a book on the subject and implemented from scratch a dictionary learning toolbox for Matlab.

pirofti commented:
Is this MRG request dead? What should be done to revive it?

adrinjalali commented:
I don't think we're much maintaining the dict learning part of the lib these days, and it's gonna be hard to find reviewers here. So closing this PR, unless at least two of the @scikit-learn/core-devs  are happy to add this in.
